Rewriting ECC

vincenthz opened the pull request.

vincenthz committed:
[ECC] Improve the code base to allow multiples different implementations

* Use TypeFamilies; need to see what to do for older GHC versions
* Start implementing some API related to ECIES

vincenthz committed:
improve description

vincenthz committed:
fix imports on older versions

vincenthz committed:
add a note about scalarInverse

kazu-yamamoto committed:
adding toByteString and fromByteString to PRK.

kazu-yamamoto committed:
implementing ecdh fpr P256 and P521.

kazu-yamamoto committed:
adding Curve_P384R1.

kazu-yamamoto committed:
adding Curve_X25519.

kazu-yamamoto committed:
Eq and Show for Point and Scalar.

kazu-yamamoto committed:
implmenting encodePoint and decodePoint for TLS.

kazu-yamamoto committed:
Dropping Show from PRK.

kazu-yamamoto committed:
removing a trailing space / a warning.

kazu-yamamoto committed:
using ByteArray(Access) instead of ByteString.

kazu-yamamoto committed:
using ScrubbedBytes directly.

kazu-yamamoto committed:
documentation & relaxing types.

kazu-yamamoto committed:
making PRK an instance of ByteArrayAccess and removing fromPRK/toPRK.

kazu-yamamoto committed:
relaxing types of encodePoint and decodePoint.

vincenthz committed:
don't export function that replace existing functionality and by-pass errors handling

vincenthz committed:
add new EC errors

vincenthz committed:
fix Curve25519 generate secret key to work in the MonadRandom instead of IO

vincenthz committed:
change point decoding to be able to fail explicitely instead of async error call.

vincenthz committed:
remove arithmetic on Curve25519. it's mathematically not possible

vincenthz committed:
fixup haddock markup

vincenthz committed:
Rewrite EC primitive and types to have the curve as type

ocheron commented on Crypto/ECC/Simple/Prim.hs:
```diff
@@ -0,0 +1,197 @@
+-- | Elliptic curve double-scalar multiplication (uses Shamir's trick).
+--
+-- > pointAddTwoMuls c n1 p1 n2 p2 == pointAdd c (pointMul c n1 p1)
+-- >                                             (pointMul c n2 p2)
```
In the equation you can remove the argument `c`.

vincenthz replied:
Thanks for pointing it out. done

vincenthz committed:
remove reference to the old api in the documentation

vincenthz committed:
add the binding to get the size by bytes

vincenthz committed:
add internal proxy type to create witnesses

vincenthz committed:
complete rewrite of the type class

Now there's no type created by associated type, it just become a routing type class,
however this has a cost, since the associated type are not injective,
requiring more witness for the curve than before.

vincenthz committed:
fix ECIES to work with the rewrite

vincenthz committed:
add some documentation to ECIES

vincenthz committed:
properly check for point validity before making a point

vincenthz committed:
use binary serializer for P256 instead of going through the simple point layer

vincenthz committed:
make a faster and more secure related to memory blits of pointDh for P256

vincenthz committed:
Add missing compatibility modules

vincenthz committed:
use the correct compat imports

vincenthz committed:
fix proxy
